codec-haproxy/src/main/java/io/netty/handler/codec/haproxy/HAProxyMessageDecoder.java

```diff
@@ -357,4 +357,36 @@ private void fail(final ChannelHandlerContext ctx, String errMsg, Throwable t) {
         }
         throw ppex;
     }
+
+    /**
+     * Returns {@code true} if the given {@link ByteBuf} is a valid HAProxy header.
+     * Be aware that this method will not increase the readerIndex of the given {@link ByteBuf}.
+     *
+     * @param buffer
+     *                  The {@link ByteBuf} to read from. Be aware that it must have at least 12 bytes to read,
+     *                  otherwise it will throw an {@link IllegalArgumentException}.
+     * @return isValid
+     *                  {@code true} if the {@link ByteBuf} is a valid HAProxy header, {@code false} otherwise.
+     * @throws IllegalArgumentException
+     *                  Is thrown if the given {@link ByteBuf} has not at least 12 bytes to read.
+     */
+    public static boolean isValidHeader(final ByteBuf buffer) {
```

Comment:
Hmm.. I feel like we need more general type to provide protocol detection mechanism in a more useful way.  For example:

```java
public interface ProtocolDetectionResult<T> { // T is HAProxyProtocolVersion or HttpVersion
    boolean needsMore();
    boolean isInvalid();
    boolean isDetected(); // or use an enum
    T detectedProtocol()
}
...

ProtocolDetectionResult<HAProxyVersion> detectProtocol(ByteBuf buf) { ... }
```